auth/grpc/auth.go

```diff
@@ -0,0 +1,173 @@
+// attached to it.
+func NewServerAuthorizer(authVerifier *token.Auth, permissions map[string][]string) (grpc.UnaryServerInterceptor, error) {
+	s := ServerAuthorizer{
+		permissions:  permissions,
+		authVerifier: authVerifier,
+	}
+	return s.Interceptor, nil
+}
+
+// Interceptor checks the provided tokens and either returns an error that includes the required
+// token scope and actions, or allows the request to proceed
+// TODO: are the error responses the ones we want to use
+func (s ServerAuthorizer) Interceptor(ctx context.Context, req interface{}, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (resp interface{}, err error) {
+	if s.authVerifier != nil {
+		gnr, ok := req.(guner)
+		if !ok {
+			return &google_protobuf.Empty{}, grpc.Errorf(
+				codes.Unauthenticated,
+				"no authorization credentials provided",
+			)
+		}
+		md, ok := metadata.FromContext(ctx)
+		var rawToken string
+		if ok {
+			ts := md["authorization"]
+			if len(ts) > 0 {
+				rawToken = ts[0]
+			}
+		}
+		if _, err := s.authVerifier.Authorize(rawToken); !ok || err != nil {
+			md := s.authVerifier.ChallengeHeaders(
+				err,
+				token.BuildAccessRecords(
+					gnr.GetGun(),
+					s.permissions[info.FullMethod]...,
+				)...,
+			)
+			grpc.SendHeader(ctx, md)
+			return &google_protobuf.Empty{}, grpc.Errorf(
+				codes.Unauthenticated,
+				"no authorization credentials provided",
+			)
+		}
+	}
+	return handler(ctx, req)
+}
+
+// ClientAuthorizer deals with satisfying tokens required by the server. If it receives an
+// error response, it will attempt to retrieve a token the server will accept
+type ClientAuthorizer struct {
+	authHandler *auth.TokenHandler
+}
+
+func NewClientAuthorizer() grpc.UnaryClientInterceptor {
+	c := ClientAuthorizer{
+		authHandler: auth.NewTokenHandler(
+			http.DefaultTransport,
+			credStore{},
+			"registry-client",
+			"",
+		),
+	}
+	return c.Interceptor
+}
+
+// Interceptor attempts to retrieve and attach the appropriate tokens for the request
+// being made
+func (c *ClientAuthorizer) Interceptor(ctx context.Context, method string, req, reply interface{}, cc *grpc.ClientConn, invoker grpc.UnaryInvoker, opts ...grpc.CallOption) error {
+	headers := metadata.MD{}
+	opts = append(opts, grpc.Header(&headers))
+	err := invoker(ctx, method, req, reply, cc, opts...)
+	if err == nil {
+		// no error, we can immediately return
+		return nil
+	}
+
+	logrus.Error(err)
+	code := grpc.Code(err)
+	if code != codes.Unauthenticated {
+		// an error other than unauthenticated, there's nothing we can do further to try
+		// and make this request succeed.
+		return err
+	}
+
+	tok, errToken := c.getToken(headers["www-authenticate"])
+	if errToken != nil {
+		// couldn't get a token, log the error and return the original Unauthenticated error
+		// (the caller of the GRPC method may be relying on a grpc type error)
+		logrus.Error(err)
+		return err
+	}
+	logrus.Info("token")
```

Comment:
Totally :-P Quite a few were me debugging because Gogland truncates how much of a value you can view in its debugger.